### PR TITLE
refactor: seclude request from aws provider

### PR DESCRIPTION
**Pull request by AlinoeDoctari**
Closes: #8445

### CODE REVIEW

**Comment by pgrzesik** [comment]
Thank you @AlinoeDoctari - @medikoo, would you be able to review this one? It's a rather big refactor of core functionality and seeing that you've specified the issue in detail, I feel like you're better equipped to properly review this PR.

File: lib/plugins/aws/deploy/lib/checkForChanges.js [hunk 1]
```diff
@@ -187,7 +187,6 @@ module.exports = {
    */
   async checkLogGroupSubscriptionFilterResourceLimitExceeded() {
     const region = this.provider.getRegion();
-    const cloudWatchLogsSdk = new this.provider.sdk.CloudWatchLogs({ region });
```
Comment (medikoo) [inline lib/plugins/aws/deploy/lib/checkForChanges.js:190]: Ideally if this PR only covers only `request`util seclucion. Can we have changes as made in this file proposed with other PR ?

File: lib/plugins/aws/utils/request.js [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+
+const MAX_RETRIES = (() => {
+  const userValue = Number(process.env.SLS_AWS_REQUEST_MAX_RETRIES);
+  return userValue >= 0 ? userValue : 4;
+})();
+
+function awsRequest(service, method, params = {}, serviceOptions = {}) {
+  // Checks regarding expectations on serviceOptions
+  if (!_.isObject(serviceOptions) || !_.isObject(serviceOptions.credentials)) {
+    return BbPromise.reject(
```
Comment (medikoo) [inline lib/plugins/aws/utils/request.js:83]: Let's not use `BbPromise`. It should be native promises and async/await syntax only

File: lib/plugins/aws/utils/request.js [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+              reject(e);
+            }
+          });
+      };
+      return doCall(0);
+    });
+
+  // Support S3 Transfer Acceleration
+  if (
+    canUseS3TransferAcceleration(service, method) &&
```
Comment (medikoo) [inline lib/plugins/aws/utils/request.js:131]: Let's remove this function, and put this resolution inline (it's one time used one liner, I don't think it's justified to have it constructed with extra function)

File: lib/plugins/aws/utils/request.js [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+    serviceOptions.isS3TransferAccelerationEnabled
+  ) {
+    enableS3TransferAcceleration(serviceOptions.credentials);
```
Comment (medikoo) [inline lib/plugins/aws/utils/request.js:134]: Same with that one

File: test/README.md [hunk 5]
```diff
@@ -50,6 +50,8 @@ AWS_ACCESS_KEY_ID=XXX AWS_SECRET_ACCESS_KEY=xxx npm run integration-test-run-all
 
 _Note: Home folder is mocked for test run, therefore relying on `AWS_PROFILE` won't work. _ and _secret key_, need to be configured directly into env variables\_
 
+\_Note: This run test depending on shared infrastructure stack (see below)
```
Comment (medikoo) [inline test/README.md:53]: Can we put it with different PR (PR's are merged with single commits and ideally if no side, not relevant changes are included with the commit) ?

Let's fix the formatting, and reword it to: _Some integration tests depend on shared infrastructure stack (see below)_

File: lib/plugins/aws/utils/request.js [hunk 6]
```diff
@@ -0,0 +1,208 @@
+'use strict';
```
Comment (medikoo) [inline lib/plugins/aws/utils/request.js:1]: Let's put it into `lib/aws/request.js`

File: lib/plugins/aws/utils/request.js [hunk 7]
```diff
@@ -0,0 +1,208 @@
...
+}
+PromiseQueue.configure(BbPromise.Promise);
+const requestQueue = new PromiseQueue(2, Infinity);
+
+const MAX_RETRIES = (() => {
+  const userValue = Number(process.env.SLS_AWS_REQUEST_MAX_RETRIES);
+  return userValue >= 0 ? userValue : 4;
+})();
+
+function awsRequest(service, method, params = {}, serviceOptions = {}) {
```
Comment (medikoo) [inline lib/plugins/aws/utils/request.js:80]: I know it was not outlined in issue description, but as I look into it now. I think it'll be better if if `service` and `serviceOptions` arguments are cumulated into one.

Same as we do it here: https://github.com/serverless/test/blob/bd685dc41a35c39da9d82ccd6fc5d60ec5d4be62/aws-request.js#L21 (note also `getServiceInstance` method).

File: lib/plugins/aws/utils/request.js [hunk 8]
```diff
@@ -0,0 +1,208 @@
...
+  // TODO enable more S3 APIs?
+  return service === 'S3' && ['upload', 'putObject'].indexOf(method) !== -1;
+}
+
+function enableS3TransferAcceleration(credentials) {
+  log('Using S3 Transfer Acceleration Endpoint...');
+  credentials.useAccelerateEndpoint = true; // eslint-disable-line no-param-reassign
+}
+
+awsRequest.memoized = memoize(awsRequest);
```
Comment (medikoo) [inline lib/plugins/aws/utils/request.js:206]: There are few issues with this memoization setup:
1.  By default `memoizee` will memoize arguments by _instance_ and not by content.
2. `memoizee` by default memoizes only against required function arguments. 
3. Async functions need to be communicated with `promise: true` option

So configuration as put cannot work well.

To fix it, the only way is to provide `normalizer` option (which technically is about resolving a cache key) as below:

```javascript
const deepSortObjectByKey = require('../utils/deepSortObjectByKey');

awsRequest.memoized = memoize(awsRequest, { promise: true,  normalizer: ([serviceData, method, params = {}]) => {
  if (typeof serviceData === 'string') serviceData = { name: serviceData });
  return [deepSortObjectByKey(serviceData), method, deepSortObjectByKey(params)].join("|");
}) });
```

**Review by medikoo — CHANGES_REQUESTED** [review]
Thank you @AlinoeDoctari ! In general it looks very good, I have just few improvement suggstions. Please see my comments

File: lib/plugins/aws/deploy/lib/checkForChanges.js (hunk 1 above)
Comment (AlinoeDoctari) [reply lib/plugins/aws/deploy/lib/checkForChanges.js:190]: Yes, no problem with moving that. Do we want to merge this other PR first or should I reintroduce the sdk member as part of the provider?

File: lib/plugins/aws/utils/request.js (hunk 2 above)
Comment (AlinoeDoctari) [reply lib/plugins/aws/utils/request.js:83]: Yes, I read that in the guidelines afterward :disappointed: I will change it :smile:

File: lib/plugins/aws/utils/request.js (hunk 7 above)
Comment (AlinoeDoctari) [reply lib/plugins/aws/utils/request.js:80]: ok, I will change it to match this implementation. It looks better indeed.

File: lib/plugins/aws/utils/request.js (hunk 8 above)
Comment (AlinoeDoctari) [reply lib/plugins/aws/utils/request.js:206]: Ok, I need to have a deeper look into it :smile: I definitely missed a couple things there.

File: lib/plugins/aws/deploy/lib/checkForChanges.js (hunk 1 above)
Comment (medikoo) [reply lib/plugins/aws/deploy/lib/checkForChanges.js:190]: I think it'll be better to first clean this up

Comment (AlinoeDoctari) [reply lib/plugins/aws/deploy/lib/checkForChanges.js:190]: I've opened the new PR here: https://github.com/serverless/serverless/pull/8857. I will put this one as WIP while we close it (I need to at least fix the Bbpromises there).

Comment (medikoo) [reply lib/plugins/aws/deploy/lib/checkForChanges.js:190]: Great. Thank you!

File: test/README.md (hunk 5 above)
Comment (AlinoeDoctari) [reply test/README.md]: moved to #8867

Comment (AlinoeDoctari) [reply test/README.md]: This has been merged.

File: lib/plugins/aws/utils/request.js (hunk 2 above)
Comment (AlinoeDoctari) [reply lib/plugins/aws/utils/request.js]: No more BbPromises in aws request and aws request test.

File: lib/plugins/aws/utils/request.js (hunk 8 above)
Comment (AlinoeDoctari) [reply lib/plugins/aws/utils/request.js]: I changed it to match yours but using the new signature. I think the behavior should be ok now.

File: lib/plugins/aws/utils/request.js (hunk 6 above)
Comment (AlinoeDoctari) [reply lib/plugins/aws/utils/request.js:1]: I'm not sure about the path here, is it `serverless/lib/aws` or `serverless/lib/plugins/aws/lib`? Because the first option would mean a new directory and it would also break the plugin isolation a bit.

Comment (medikoo) [reply lib/plugins/aws/utils/request.js:1]: > Because the first option would mean a new directory and it would also break the plugin isolation a bit.

It's `lib/aws/request`. Note it's not a plugin anymore, also everything in `lib` should be considered private

Comment (AlinoeDoctari) [reply lib/plugins/aws/utils/request.js:1]: Ok :smile: will do :smile: thanks. Just to try and understand the big plan here, is there a plan to move aws outside of the plugin dir?

Comment (medikoo) [reply lib/plugins/aws/utils/request.js:1]: Utils will rather be moved here, but it's an ongoing progress. It might be even that some point Framework will get modularized and AWS stuff will be moved to other repository (but that's a long term consideration, nothing we will rush with now)

Comment (AlinoeDoctari) [reply lib/plugins/aws/utils/request.js]: Thanks for the answer :smile: I've moved it there in the last commit.

File: lib/plugins/aws/deploy/lib/checkForChanges.js (hunk 1 above)
Comment (AlinoeDoctari) [reply lib/plugins/aws/deploy/lib/checkForChanges.js]: Rebased on master after closure of #8857 . Should be ok to work on it again :smile:

**Comment by AlinoeDoctari** [comment]
Squashed and rebased on the latest version.

File: lib/aws/request.js [hunk 9]
```diff
@@ -0,0 +1,234 @@
...
+        useAccelerateEndpoint: shouldS3Accelerate(),
+        region: service.params.region ? service.params.region : 'us-east-1',
+      };
+      return new Service(serviceParams);
+    }
+    const serviceParams = {
+      credentials: service.params.credentials,
+      region: service.params.region ? service.params.region : 'us-east-1',
+    };
+    return new Service(serviceParams);
```
Comment (medikoo) [inline lib/aws/request.js:117]: Maintanance wise, it'll be nicer to do:

```javascript
const serviceParams = {
  credentials: ..
  region: ...
}
if (service.name === 'S3') serviceParams.useAccelerateEndpoint = ...
```

**Comment by AlinoeDoctari** [comment]
@medikoo Should I reintroduce Bbpromises after #9074?

**Comment by medikoo** [comment]
>  @medikoo Should I reintroduce Bbpromises after #9074?

No, definitely not. I'm reviewing it just right now. Sorry for late review, first I didn't notice it was converted from draft (that doesn't generate real notification)

**Comment by medikoo** [comment]
@AlinoeDoctari ideally, give me a moment to finalize the review, then it'll be good if you fix the conflicts and address the comments.

Great thanks with help on that!

**Comment by AlinoeDoctari** [comment]
Yes, no problem, take your time :+1:

File: lib/aws/request.js [hunk 10]
```diff
@@ -0,0 +1,234 @@
...
+                ].join(' ')
+              );
+              setTimeout(doCall, backOff, nextTryNum);
+            } else {
+              reject(e);
+            }
+          });
+      };
+      return doCall(0);
+    });
```
Comment (medikoo) [inline lib/aws/request.js:174]: Let's construct this as:

```javascript
const persistentRequest = (f, numTry = 0) => f().catch(e => {
  ...
  return setTimeout(persistentRequest, backOff, f, nextNumTry);
});
```

Current promise construct is leaky and shares bad patterns

File: lib/aws/request.js [hunk 11]
```diff
@@ -0,0 +1,234 @@
...
+          const errorMessage = bottomError.message.startsWith('EC2 Metadata')
+            ? [
+                'AWS provider credentials not found.',
+                ' Learn how to set up AWS provider credentials',
+                ` in our docs here: <${chalk.green('http://slss.io/aws-creds-setup')}>.`,
+              ].join('')
+            : bottomError.message;
+          message = errorMessage;
+          // We do not want to trigger the retry mechanism for credential errors
+          return Promise.reject(
```
Comment (medikoo) [inline lib/aws/request.js:206]: Let's simply `throw` here

File: lib/aws/request.js [hunk 12]
```diff
@@ -0,0 +1,234 @@
...
+            Object.assign(new ServerlessError(errorMessage, err.code), {
+              providerError: Object.assign({}, err, { retryable: false }),
+            })
+          );
+        }
+
+        return Promise.reject(
```
Comment (medikoo) [inline lib/aws/request.js:213]: Same here

File: lib/aws/request.js [hunk 13]
```diff
@@ -0,0 +1,234 @@
...
+          Object.assign(new ServerlessError(message, err.code), {
+            providerError: err,
+          })
+        );
+      });
+    }).then((data) => {
+      const result = Promise.resolve(data);
+      return result;
+    })
```
Comment (medikoo) [inline lib/aws/request.js:222]: Let's remove this handler, in given form it doesn't do anything

File: lib/aws/request.js [hunk 14]
```diff
@@ -0,0 +1,234 @@
...
+  );
+
+  return request;
+}
+
+awsRequest.memoized = memoize(awsRequest, {
+  normalizer: ([service, method, args]) => {
+    return [deepSortObjectByKey(service), method, deepSortObjectByKey(args)].join('|');
```
Comment (medikoo) [inline lib/aws/request.js:230]: We need to stringify arguments (Possibly best would be to `[].map(value => JSON.stringify(value)).join("|")`). Also I think `service` can be a string (?)

File: lib/aws/request.js [hunk 15]
```diff
@@ -0,0 +1,234 @@
...
+})();
+
+const getServiceInstance = memoize(
+  (service, method) => {
+    // Checks regarding expectations on service object
+    if (
+      !_.isObject(service) ||
+      !_.isObject(service.params) ||
+      !_.isString(service.name) ||
+      !_.isObject(service.params.credentials)
```
Comment (medikoo) [inline lib/aws/request.js:87]: Let's validate this with [type](https://github.com/medikoo/type#readme) library. We already do it in some places, e.g. https://github.com/serverless/serverless/blob/b81fd0557f53e11597fe67bb138c09675d61e0c5/lib/Serverless.js#L33-L44

**Review by medikoo — CHANGES_REQUESTED** [review]
Thank you @AlinoeDoctari ! It looks very promising, please see my comments

File: lib/aws/request.js (hunk 10 above)
Comment (AlinoeDoctari) [reply lib/aws/request.js:174]: Not sure about this one, I've tried an implementation, let me know if it matches what you had in mind :slightly_smiling_face:

File: lib/aws/request.js (hunk 14 above)
Comment (AlinoeDoctari) [reply lib/aws/request.js:230]: At that point, not anymore, I've moved the type assertions to make it clearer.

File: lib/aws/request.js (hunk 15 above)
Comment (AlinoeDoctari) [reply lib/aws/request.js:87]: It should definitely make the error message more meaningful.

File: lib/aws/request.js (hunk 14 above)
Comment (AlinoeDoctari) [reply lib/aws/request.js:230]: After checking a bit, this is also the case for the args, so I've modified the implementation to include the stringify.

**Comment by AlinoeDoctari** [comment]
@medikoo I've not rebased yet to try to help the review of the changes. Just let me know when you want me to do it to remove the conflicts.

**Comment by medikoo** [comment]
> @medikoo I've not rebased yet to try to help the review of the changes. Just let me know when you want me to do it to remove the conflicts.

@AlinoeDoctari please rebase, that will also run CI, which will confirm on issues that can be noticed automatically.
After that's done and you feel it's ready for review, please re-request it in top right section (that triggers a notification for me, which I usually promptly follow :)

File: lib/aws/request.js [hunk 16]
```diff
@@ -0,0 +1,218 @@
...
+ * @prop [service.name] - Name of the service to call, support subclasses
+ * @prop [service.params] - Parameters to apply when creating the service and doing the request
+ * @prop [service.params.credentials] - AWS Credentials to use
+ * @prop [service.params.useCache ] - Wether to reuse result of the same request cached locally
+ * @prop [service.params.region] - Region in which the call should be made (default to us-east-1)
+ * @prop [service.params.isS3TransferAccelerationEnabled] - Use s3 acceleration when available for the request
+ * @param {String} method - Method to call
+ * @param {Array} args - Argument for the method call
+ */
+function awsRequest(service, method, ...args) {
```
Comment (medikoo) [inline lib/aws/request.js:126]: This should be async function

File: lib/aws/request.js [hunk 17]
```diff
@@ -0,0 +1,218 @@
...
+        });
+      });
+    })
+  );
+
+  return request;
+}
+
+awsRequest.memoized = memoize(awsRequest, {
+  normalizer: ([service, method, args]) => {
```
Comment (medikoo) [inline lib/aws/request.js:209]: Let's also add `promise: true`, as we don't want to memoize rejections

**Review by medikoo — CHANGES_REQUESTED** [review]
@AlinoeDoctari thank you, looks great! I've suggested few more improvements, and I think we're getting close to having it in :)

File: lib/aws/request.js [hunk 18]
```diff
@@ -0,0 +1,218 @@
...
+  ensureObject(service, { name: 'service', Error: ServerlessError });
+  ensureObject(service.params, { name: 'service.params', Error: ServerlessError });
+  ensureObject(service.params.credentials, {
+    name: 'service.params.credentials',
+    Error: ServerlessError,
+  });
+  ensureString(service.name, { name: 'service.name', Error: ServerlessError });
+  const BASE_BACKOFF = 5000;
+  const persistentRequest = (f, numTry = 0) =>
+    f().catch((e) => {
```
[the remark this answers is not in the document]
Comment (AlinoeDoctari) [reply lib/aws/request.js:137]: Okay :smile: I hesitated since the changes where more pervasive but fine by me :smile:

File: lib/aws/request.js [hunk 19]
```diff
@@ -0,0 +1,218 @@
...
+ */
+function awsRequest(service, method, ...args) {
+  // Checks regarding expectations on service object
+  ensureObject(service, { name: 'service', Error: ServerlessError });
+  ensureObject(service.params, { name: 'service.params', Error: ServerlessError });
+  ensureObject(service.params.credentials, {
+    name: 'service.params.credentials',
+    Error: ServerlessError,
+  });
+  ensureString(service.name, { name: 'service.name', Error: ServerlessError });
```
[the remark this answers is not in the document]
Comment (AlinoeDoctari) [reply lib/aws/request.js:134]: Does that mean that we should support making request without authentication?

[the remark this answers is not in the document]
Comment (medikoo) [reply lib/aws/request.js:134]: If not provided with params, AWS SDK reads authentication on its own internally reading from env vars.

I think it's nice to leave an option for this util to be used that way. (by secluding, it's no longer tight to `Serverless` instance, so it could be used e.g. by scripts)

Also it's possible that in context of a Framework we will refactor internal credentials handling, so it'll be not passed with params, but set directly on `process.env` (to be discussed)

[the remark this answers is not in the document]
Comment (AlinoeDoctari) [reply lib/aws/request.js:134]: Ok, this is a bigger discussion indeed :smile:  I'll remove the constraint there and try to support the service/string use case with associated test.

[the remark this answers is not in the document]
Comment (AlinoeDoctari) [reply lib/aws/request.js]: Unfortunately this lead me to remove most of the authentication checks. The failure were happening when creating the service and not later on. I'd like to write the authentication scenario and failure mode for that awsRequest function if we want to introduce a bit more checking.

[the remark this answers is not in the document]
Comment (medikoo) [reply lib/aws/request.js]: > Unfortunately this lead me to remove most of the authentication checks.

Do you mean checks that confirm that those options are validated. Or some other  checks?

[the remark this answers is not in the document]
Comment (AlinoeDoctari) [reply lib/aws/request.js]: Other checks, including the one redirecting to serverless documentation for authentication issues.

**Review by medikoo — CHANGES_REQUESTED** [review]
Thank you @AlinoeDoctari ! It looks really good now, wee're getting close, please see my comments

File: lib/aws/request.js (hunk 19 above)
[the remark this answers is not in the document]
Comment (medikoo) [reply lib/aws/request.js]: I've just posted review, I believe we really should keep that part. What was the reason you've decided to strip it?

I think all it does, is making an authentication error communicated with way better message, than if we surface one from AWS directly

File: lib/aws/request.js [hunk 20]
```diff
@@ -0,0 +1,198 @@
...
+ * @prop [service.params.useCache ] - Wether to reuse result of the same request cached locally
+ * @prop [service.params.region] - Region in which the call should be made (default to us-east-1)
+ * @prop [service.params.isS3TransferAccelerationEnabled] - Use s3 acceleration when available for the request
+ * @param {String} method - Method to call
+ * @param {Array} args - Argument for the method call
+ */
+async function awsRequest(service, method, ...args) {
+  // Checks regarding expectations on service object
+  if (isObject(service)) {
+    ensureString(service.name, { name: 'service.name', Error: ServerlessError });
```
[the remark this answers is not in the document]
Comment (AlinoeDoctari) [reply lib/aws/request.js:133]: Indeed, I'll remove that.

File: lib/aws/request.js [hunk 21]
```diff
@@ -0,0 +1,198 @@
...
+    }
+  };
+  const request = await requestQueue.add(() =>
+    persistentRequest(async () => {
+      const awsService = getServiceInstance(service, method);
+      const req = awsService[method](...args);
+      try {
+        return await req.promise();
+      } catch (err) {
+        const message = err.message != null ? err.message : String(err.code);
```
[the remark this answers is not in the document]
Comment (AlinoeDoctari) [reply lib/aws/request.js:177]: I'll try to describe the authentication options supported for that function in the test and reintroduce the it. This appeared broken when trying to use the environment variable credentials.

File: lib/aws/request.js [hunk 22]
```diff
@@ -0,0 +1,224 @@
...
+
+const getServiceInstance = memoize(
+  (service, method) => {
+    const Service = _.get(sdk, service.name);
+    // we translate params to an object for the service creation by selecting keys of interest
+    const serviceParams = {};
+    if (service.params) {
+      if (service.params.credentials) serviceParams.credentials = service.params.credentials;
+      if (service.params.region) serviceParams.region = service.params.region;
+    }
```
Comment (medikoo) [inline lib/aws/request.js:105]: Maybe let's simply do `const serviceParams = { ...service.params };`

I think it'll be more versatile, than picking params one by one by name

File: lib/aws/request.js [hunk 23]
```diff
@@ -0,0 +1,224 @@
...
+    if (service.name === 'S3') {
+      serviceParams.useAccelerateEndpoint = shouldS3Accelerate(method, service.params);
+    }
+    if (_.isEmpty(serviceParams)) {
+      return new Service();
+    }
+    return new Service(serviceParams);
```
Comment (medikoo) [inline lib/aws/request.js:112]: I gues both calls are equivalent, so it can be just `return new Service(params)`

File: test/unit/lib/aws/request.test.js [hunk 24]
```diff
@@ -0,0 +1,401 @@
+'use strict';
```
Comment (medikoo) [inline test/unit/lib/aws/request.test.js:1]: This should be located in `test/unit/aws/request.test.js`

**Review by medikoo — CHANGES_REQUESTED** [review]
@AlinoeDoctari that looks great! I have pointed just few last mostly style related bits, and we should be good to go

File: lib/aws/request.js (hunk 22 above)
Comment (AlinoeDoctari) [reply lib/aws/request.js:105]: Definitely, thank you :+1:

File: test/unit/lib/aws/request.test.js (hunk 24 above)
Comment (AlinoeDoctari) [reply test/unit/lib/aws/request.test.js:1]: Yes, it should :smile:

Comment (AlinoeDoctari) [reply test/unit/lib/aws/request.test.js:1]: Actually this will not match the current organization in the test/unit folder right? We currently have the exact same tree in `test/unit/lib` and `lib` after that change that won't be the case anymore. @medikoo is that expected?

Comment (medikoo) [reply test/unit/lib/aws/request.test.js]: ah... sorry, ignore that comment! I got confused :) All is good

**Comment by AlinoeDoctari** [comment]
@medikoo Should be all good for now :slightly_smiling_face:

File: lib/aws/request.js [hunk 25]
```diff
@@ -0,0 +1,217 @@
...
+    })
+  );
+  return request;
+}
+
+awsRequest.memoized = memoize(awsRequest, {
+  promise: true,
+  normalizer: ([service, method, args]) => {
+    return [
+      JSON.stringify(deepSortObjectByKey(service)),
```
Comment (medikoo) [inline lib/aws/request.js:210]: I think `service `could be string here. In such case it'll be nice to convert it to `{ name: service }`

**Review by medikoo — CHANGES_REQUESTED** [review]
@AlinoeDoctari looks great, I've just spotted one I believe last issue, related to memoized version, please see my comment

File: lib/aws/request.js (hunk 25 above)
Comment (AlinoeDoctari) [reply lib/aws/request.js:210]: That's right nice catch :+1:

File: lib/aws/request.js [hunk 26]
```diff
@@ -0,0 +1,218 @@
...
+      }
+    })
+  );
+  return request;
+}
+
+awsRequest.memoized = memoize(awsRequest, {
+  promise: true,
+  normalizer: ([service, method, args]) => {
+    if (_.isString(service)) service = { name: service };
```
Comment (medikoo) [inline lib/aws/request.js:209]: It would be good to follow same resolution as in memoized function (and also we do not use lodash if not justified, e.g. in this case it could be `typeof service === 'string'`).

Anyway if we want to reflect what happens in `request`. It'll be:

```javascript
if (!isObject(service)) service = { name: ensureString(service) }
```

**Review by medikoo — CHANGES_REQUESTED** [review]

File: lib/aws/request.js (hunk 26 above)
Comment (AlinoeDoctari) [reply lib/aws/request.js:209]: I'll change it accordingly, I wanted to avoid having an error in the memoized part and keep it for the actual call.

Comment (medikoo) [reply lib/aws/request.js:209]: It's totally fine to reject here. It'll have exactly same effect

**Review by medikoo — APPROVED** [review]
Thank you @AlinoeDoctari ! Great work
